pkg/frontend/fixetcd.go

```diff
@@ -0,0 +1,669 @@
+}
+
+// comparePodEnvToIp compares the etcd container's environment variables to the pod's actual IP address
+func comparePodEnvToIP(log *logrus.Entry, pods *corev1.PodList) (*degradedEtcd, error) {
+	de := &degradedEtcd{}
+	for _, p := range pods.Items {
+		etcdIP := ipFromEnv(p.Spec.Containers, p.Name)
+		for _, ip := range p.Status.PodIPs {
+			if ip.IP != etcdIP && etcdIP != "" {
+				log.Infof("Found conflicting IPs for etcd Pod %s: %s!=%s", p.Name, ip.IP, etcdIP)
+				de.Node = strings.ReplaceAll(p.Name, "etcd-", "")
+				de.Pod = p.Name
+				de.NewIP = ip.IP
+				de.OldIP = etcdIP
+			}
+		}
+	}
+
+	// If no conflict is found a recent IP change may still be causing an issue
+	// Sometimes etcd can recovery the deployment itself, however there is still a data directory with the previous member's IP address present causing a failure
+	// This can still be remediated by relying on the pod statuses
+	if de.Node == "" {
+		log.Info("Unable to find an IP address conflict, checking etcd container statuses")
+		crashingPod, err := findCrashloopingPods(log, pods)
+		if err != nil {
+			return nil, err
+		}
+
+		de.Node = strings.ReplaceAll(crashingPod.Name, "etcd-", "")
+		de.Pod = crashingPod.Name
+		de.NewIP = "unknown"
+		de.OldIP = "unknown"
+		log.Infof("Found %s, with condition ready false: %v", de.Pod, de)
+	}
+
+	return de, nil
+}
+
+func ipFromEnv(containers []corev1.Container, podName string) string {
+	for _, c := range containers {
+		if c.Name == "etcd" {
+			for _, e := range c.Env {
+				// The environment variable that contains etcd's IP address has the following naming convention
+				// NODE_cluster_name_c4j2v_master_0_IP
+				// while the pod looks like this
+				// etcd-cluster-name-c4j2v-master-0
+				// To find the pod's IP address by variable name we use the pod's name
+				envName := strings.ReplaceAll(strings.ReplaceAll(podName, "-", "_"), "etcd_", "NODE_")
+				if e.Name == fmt.Sprintf("%s_IP", envName) {
+					return e.Value
+				}
+			}
+		}
+	}
+	return ""
+}
+
+// TODO make this work for non status checks too
+func findCrashloopingPods(log *logrus.Entry, pods *corev1.PodList) (*corev1.Pod, error) {
+	// pods are collected in a list to check for multiple crashing etcd instances
+	// multiple etcd failures aren't supported so an error will be returned, rather than assuming the first found is the only one
+	crashingPods := &corev1.PodList{}
+	for _, p := range pods.Items {
+		for _, c := range p.Status.ContainerStatuses {
+			if !c.Ready && c.Name == "etcd" {
+				log.Infof("Found etcd container %s not ready", c.ContainerID)
+				crashingPods.Items = append(crashingPods.Items, p)
+			}
+		}
+	}
+
+	if len(crashingPods.Items) > 1 {
+		// log multiple names in a readable way
+		names := []string{}
+		for _, c := range crashingPods.Items {
+			names = append(names, c.Name)
+		}
+		return nil, fmt.Errorf("only a single degraded quorum is supported, more than one not Ready etcd pods were found: %v", names)
```

Comment:
Nit: IIUC the word "quorum" refers only to the group and can't be used to refer to individual members. I would replace "quorum" here with "etcd pod".

Reply:
@kimorris27 That makes sense to me given the definition of [quorum](https://www.merriam-webster.com/dictionary/quorum), I'll update the log message.